### PR TITLE
[JAV-508] & [SCB-11] Metrics Auto Publish

### PR DESCRIPTION
**Pull request by zhengyangyong**
Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] Make sure there is a [JIRA issue](https://https://servicecomb.atlassian.net/projects/JAV/issues) filed for the change (usually before you start working on it).  Trivial changes like typos do not require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
 - [x] Each commit in the pull request should have a meaningful subject line and body.
 - [x] Format the pull request title like `[JAV-XXX] Fixes bug in ApproximateQuantiles`, where you replace `JAV-XXX` with the appropriate JIRA issue.
 - [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [x] Run `mvn clean install` to make sure basic checks pass. A more thorough check will be performed on your pull request automatically.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---
If user include metrics in dependency,the metrics service would auto publish with Java Chassis Producer,also use same transport and endpoint config.
For example if user config rest address 0.0.0.0:8080 in microservice.yaml,then the metrics service will available at http://:8080/metrics

this PR include :
1.Direct metrics publish , interface is MetricsPublisher
public interface MetricsPublisher {
  RegistryMetric metrics();
  RegistryMetric metricsWithWindowTimeIndex(int windowTimeIndex);
}

2.Integration test in demo-springmvc

### CODE REVIEW

File: demo/demo-springmvc/springmvc-server/src/main/java/io/servicecomb/demo/springmvc/server/CodeFirstSpringmvc.java [hunk 1]
```diff
@@ -312,4 +328,19 @@ public String testform(HttpServletRequest request) {
...
+  public String metricsForTest() {
+    List<Monitor<?>> monitors = registry.getMetricsMonitors();
+    Map<String, String> values = new HashMap<>();
+    for (Monitor<?> monitor : monitors) {
+      values.put(monitor.getConfig().getName(), monitor.getValue().toString());
+    }
+    try {
+      return JsonUtils.writeValueAsString(values);
+    } catch (JsonProcessingException e) {
+      return "{}";
```
Comment (WillemJiang) [inline demo/demo-springmvc/springmvc-server/src/main/java/io/servicecomb/demo/springmvc/server/CodeFirstSpringmvc.java]: we should log the exception here.

File: metrics/metrics-common/src/main/java/io/servicecomb/metrics/common/MetricsPublisher.java [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.servicecomb.metrics.common;
+
+public interface MetricsPublisher {
+  RegistryMetric metrics();
+
+  RegistryMetric metricsWithWindowTimeIndex(int windowTimeIndex);
```
Comment (WillemJiang) [inline metrics/metrics-common/src/main/java/io/servicecomb/metrics/common/MetricsPublisher.java]: Please add some comments on this operation.  It's hard for the user to understand.

Comment (zhengyangyong) [reply metrics/metrics-common/src/main/java/io/servicecomb/metrics/common/MetricsPublisher.java]: Done

File: metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/publish/DefaultMetricsPublisher.java [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+  @RequestMapping(path = "/", method = RequestMethod.GET)
+  @Override
+  public RegistryMetric metrics() {
+    return dataSource.getRegistryMetric();
+  }
+
+  @RequestMapping(path = "/{windowTimeIndex}", method = RequestMethod.GET)
+  @Override
+  public RegistryMetric metricsWithWindowTimeIndex(@PathVariable(name = "windowTimeIndex") int windowTimeIndex) {
+    return dataSource.getRegistryMetric(windowTimeIndex);
```
Comment (wujimin) [inline metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/publish/DefaultMetricsPublisher.java]: should validate input parameter windowTimeIndex

File: metrics/metrics-common/src/main/java/io/servicecomb/metrics/common/MetricsPublisher.java [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+   *
+   *                 ↑ getRegistryMetric(0) will return max=200 min=100 total=300
+   *                   getRegistryMetric(1) will return max=0 min=0 total=0
+   *                             ↑ getRegistryMetric(0) will return max=300 min=400 total=700
+   *                               getRegistryMetric(1) will return max=400 min=100 total=1000
+   *
+   * @param windowTimeIndex index of getAppliedWindowTime() item
+   * @return RegistryMetric
+   */
+  RegistryMetric metricsWithWindowTimeIndex(int windowTimeIndex);
```
Comment (wujimin) [inline metrics/metrics-common/src/main/java/io/servicecomb/metrics/common/MetricsPublisher.java]: consumer do not what is index 0, what is index 1
index is not a good idea.

File: demo/demo-springmvc/springmvc-client/src/main/resources/META-INF/spring/springmvc.client.bean.xml [hunk 5]
```diff
@@ -28,4 +28,7 @@
 	<cse:rpc-reference id="controller" microservice-name="springmvc"
 		schema-id="controller" interface="io.servicecomb.demo.controller.Controller"></cse:rpc-reference>
 
+	<cse:rpc-reference id="metricsPublisher" microservice-name="springmvc"
```
Comment (wujimin) [inline demo/demo-springmvc/springmvc-client/src/main/resources/META-INF/spring/springmvc.client.bean.xml]: use annotation is simpler

File: metrics/metrics-core/pom.xml [hunk 6]
```diff
@@ -29,18 +29,27 @@
...
+    </dependency>
     <dependency>
       <groupId>io.servicecomb</groupId>
-      <artifactId>java-chassis-core</artifactId>
+      <artifactId>metrics-common</artifactId>
     </dependency>
     <dependency>
       <groupId>io.servicecomb</groupId>
-      <artifactId>foundation-metrics</artifactId>
+      <artifactId>provider-rest-common</artifactId>
```
Comment (wujimin) [inline metrics/metrics-core/pom.xml]: do not use rest-common + spring web
please use provider springmvc

File: metrics/metrics-core/pom.xml [hunk 7]
```diff
@@ -29,18 +29,27 @@
...
     </dependency>
     <dependency>
-      <groupId>com.netflix.servo</groupId>
-      <artifactId>servo-core</artifactId>
+      <groupId>org.springframework</groupId>
+      <artifactId>spring-web</artifactId>
+    </dependency>
+    <dependency>
```
Comment (wujimin) [inline metrics/metrics-core/pom.xml]: no need to declare java-chassis-core

File: metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/publish/DefaultMetricsPublisher.java (hunk 3 above)
Comment (zhengyangyong) [reply metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/publish/DefaultMetricsPublisher.java]: Ok ,it's a good suggestion

File: metrics/metrics-common/src/main/java/io/servicecomb/metrics/common/MetricsPublisher.java (hunk 4 above)
Comment (zhengyangyong) [reply metrics/metrics-common/src/main/java/io/servicecomb/metrics/common/MetricsPublisher.java]: Ok ,it's a good suggestion

File: metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/publish/DefaultMetricsPublisher.java (hunk 3 above)
Comment (zhengyangyong) [reply metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/publish/DefaultMetricsPublisher.java]: Done

File: metrics/metrics-common/src/main/java/io/servicecomb/metrics/common/MetricsPublisher.java (hunk 4 above)
Comment (zhengyangyong) [reply metrics/metrics-common/src/main/java/io/servicecomb/metrics/common/MetricsPublisher.java]: Done

File: demo/demo-springmvc/springmvc-client/src/main/resources/META-INF/spring/springmvc.client.bean.xml (hunk 5 above)
Comment (zhengyangyong) [reply demo/demo-springmvc/springmvc-client/src/main/resources/META-INF/spring/springmvc.client.bean.xml]: keep same as before...

File: metrics/metrics-core/pom.xml (hunk 6 above)
Comment (zhengyangyong) [reply metrics/metrics-core/pom.xml]: Done

File: metrics/metrics-core/pom.xml (hunk 7 above)
Comment (zhengyangyong) [reply metrics/metrics-core/pom.xml]: Done

File: metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/publish/DefaultDataSource.java [hunk 8]
```diff
@@ -17,41 +17,79 @@
...
   public RegistryMetric getRegistryMetric() {
-    return getRegistryMetric(0);
+    return getRegistryMetric(getAppliedWindowTime().get(0));
   }
 
   @Override
-  public RegistryMetric getRegistryMetric(int windowTimeIndex) {
-    return registryMonitor.toRegistryMetric(windowTimeIndex);
+  public RegistryMetric getRegistryMetric(long windowTime) {
+    if (appliedWindowTimes.containsKey(windowTime)) {
```
Comment (wujimin) [inline metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/publish/DefaultDataSource.java]: 1.contains then get, cause search two times
2.already return, no need "else"

Comment (zhengyangyong) [reply metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/publish/DefaultDataSource.java]: Done

**Pull request merged**